crates/ruff_linter/src/rules/ruff/rules/sequence_sorting.rs

```diff
@@ -0,0 +1,818 @@
+/// Utilities for sorting constant lists of string literals.
+///
+/// Examples where these are useful:
+/// - Sorting `__all__` in the global scope,
+/// - Sorting `__slots__` or `__match_args__` in a class scope
+use std::borrow::Cow;
+use std::cmp::Ordering;
+
+use ruff_python_ast as ast;
+use ruff_python_codegen::Stylist;
+use ruff_python_parser::{lexer, Mode, Tok};
+use ruff_python_trivia::leading_indentation;
+use ruff_source_file::Locator;
+use ruff_text_size::{Ranged, TextRange, TextSize};
+
+use is_macro;
+use itertools::{izip, Itertools};
+
+/// An enumeration of the various kinds of sequences for which Python has
+/// [display literals](https://docs.python.org/3/reference/expressions.html#displays-for-lists-sets-and-dictionaries).
+///
+/// (I'm aware a set isn't actually a "sequence",
+/// *but* for our purposes it's conceptually a sequence,
+/// since in terms of the AST structure it's almost identical
+/// to tuples/lists.)
+///
+/// Whereas lists, dicts and sets are always parenthesized
+/// (e.g. lists always start with `[` and end with `]`),
+/// single-line tuples *can* be unparenthesized.
+/// We keep the original AST node around for the
+/// Tuple variant so that this can be queried later.
+#[derive(Debug)]
+pub(super) enum SequenceKind<'a> {
+    List,
+    Set,
+    Tuple(&'a ast::ExprTuple),
+}
+
+impl SequenceKind<'_> {
+    fn surrounding_parens(&self, source: &str) -> (&str, &str) {
+        match self {
+            Self::List => ("[", "]"),
+            Self::Set => ("{", "}"),
+            Self::Tuple(ast_node) => {
+                if ast_node.is_parenthesized(source) {
+                    ("(", ")")
+                } else {
+                    ("", "")
+                }
+            }
+        }
+    }
+
+    fn opening_token_for_multiline_definition(&self) -> Tok {
+        match self {
+            Self::List => Tok::Lsqb,
+            Self::Set => Tok::Lbrace,
+            Self::Tuple(_) => Tok::Lpar,
+        }
+    }
+
+    fn closing_token_for_multiline_definition(&self) -> Tok {
+        match self {
+            Self::List => Tok::Rsqb,
+            Self::Set => Tok::Rbrace,
+            Self::Tuple(_) => Tok::Rpar,
+        }
+    }
+}
+
+/// An enumeration of the various kinds of
+/// [display literals](https://docs.python.org/3/reference/expressions.html#displays-for-lists-sets-and-dictionaries)
+/// Python provides for builtin containers.
+#[derive(Debug, is_macro::Is)]
+pub(super) enum DisplayKind<'a> {
+    Sequence(SequenceKind<'a>),
+    Dict { values: &'a Vec<ast::Expr> },
+}
+
+/// Create a string representing a fixed-up single-line
+/// definition of `__all__` or `__slots__` (etc.),
+/// that can be inserted into the
+/// source code as a `range_replacement` autofix.
+pub(super) fn sort_single_line_elements_sequence<F>(
+    kind: &SequenceKind,
+    elts: &[ast::Expr],
+    elements: &[&str],
+    locator: &Locator,
+    mut cmp_fn: F,
```

Comment:
Nit: Allowing arbitrary functions is one of the most flexible solution but it has a few downsides:

* Rust monomorphizes `sort_single_line_elements_sequence` everytime it is called and a different `F` type is passed. It means, Rust will duplicate your code. This isn't as much of a problem here because the function isn't too large and only used a few times. But it can be a big problem if the implementation is large. 
* This doesn't really apply here because `cmp_fn` is a very common pattern (maybe rename the function to `sort_single_line_elements_by`). However, accepting generic functions doesn't provide any addtional information to me as reader without reading different call-sites. 

An alternative is to introduce a new enum `SortKeyKind` or `Sortable` that has two variants, one for slots and one for dunder all. You can then match on the sort key kind to determine how the elements should be sorted.

Reply:
Would the enum solution require moving all the "isort-style" sorting logic from `sort_dunder_all.rs` into `sequence_sorting.rs`? I don't mind that too much if it's necessary, but conceptually I'd rather keep the isort-style sorting logic in `sort_dunder_all.rs`, since it's specific to that rule

Reply:
Yeah, it would require moving the sorting logic.